File: packages/nativescript-masonkit/src-native/mason-android/masonkit/src/main/java/org/nativescript/mason/masonkit/MasonPositioning.kt
package org.nativescript.mason.masonkit

import android.app.Activity
import android.content.Context
import android.content.ContextWrapper
import android.view.View
import android.view.ViewGroup

import org.nativescript.mason.masonkit.enums.Overflow
import org.nativescript.mason.masonkit.enums.Position

internal object MasonPositioning {

  // Never repositions children itself, unlike NativeScript's page GridLayout.
  private class FixedOverlay(context: Context) : ViewGroup(context) {
    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {}
    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
      setMeasuredDimension(MeasureSpec.getSize(widthMeasureSpec), MeasureSpec.getSize(heightMeasureSpec))
    }
  }

  private fun Context.findActivity(): Activity? {
    var ctx = this
    while (ctx is ContextWrapper) {
      if (ctx is Activity) return ctx
      ctx = ctx.baseContext
    }
    return null
  }

  private fun rootOf(node: Node): Node {
    var current = node
    while (current.parent != null) {
      current = current.parent!!
    }
    return current
  }

  // `view is Scroll` is true for plain divs too (Scroll doubles as a generic container); check overflow instead.
  private fun isScrollContainer(view: View): Boolean {
    val element = view as? Element ?: return false
    val overflow = element.style.overflow
    return overflow.x == Overflow.Scroll || overflow.x == Overflow.Auto || overflow.y == Overflow.Scroll || overflow.y == Overflow.Auto
  }

  // fixed's containing block: a full-screen overlay at the Activity's content root.
  private fun fixedHost(node: Node): ViewGroup? {
    val rootView = rootOf(node).view as? View ?: return null
    val activity = rootView.context.findActivity() ?: return null
    val decorContent = activity.findViewById<ViewGroup>(android.R.id.content) ?: return null
    (decorContent.getTag() as? FixedOverlay)?.let { return it }
    val overlay = FixedOverlay(decorContent.context)
    decorContent.addView(overlay, ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT))
    decorContent.tag = overlay
    return overlay
  }

  // Sum of left/top offsets from `view` up to `host`; scrollX/Y doesn't affect left/top, so this ignores scrolling.
  private fun accumulateOrigin(baseX: Int, baseY: Int, view: View, host: View): Pair<Int, Int> {
    var x = baseX
    var y = baseY
    var current = view.parent as? View
    while (current != null && current !== host) {
      x += current.left
      y += current.top
      current = current.parent as? View
    }
    return x to y
  }

  // Reparents to the fixed host once, then rewrites (x,y) from root-relative to the host's space.
  fun applyFixed(node: Node, view: View, x: Int, y: Int): Pair<Int, Int> {
    val host = fixedHost(node) ?: return x to y
    val rootView = rootOf(node).view as? View ?: return x to y
    if (view.parent !== host) {
      if (node.fixedOriginalParent == null) {
        node.fixedOriginalParent = view.parent as? ViewGroup
      }
      (view.parent as? ViewGroup)?.removeView(view)
      host.addView(view)
    } else {
      host.bringChildToFront(view)
    }
    val (originX, originY) = accumulateOrigin(rootView.left, rootView.top, rootView, host)
    return (originX + x) to (originY + y)
  }

  // Undoes applyFixed once a node stops being fixed.
  fun clearPositioning(node: Node, view: View) {
    node.stickyScrollHost = null
    val original = node.fixedOriginalParent
    if (original != null) {
      node.fixedOriginalParent = null
      if (view.parent !== original) {
        (view.parent as? ViewGroup)?.removeView(view)
        original.addView(view)
      }
    }
  }

  // Walks the native parent chain, not Mason's node-parent chain, which other mounted screens can corrupt.
  private fun nearestScrollAncestor(view: View): Scroll? {
    var current = view.parent as? View
    while (current != null) {
      if (isScrollContainer(current)) {
        (current as? Scroll)?.let { return it }
      }
      current = current.parent as? View
    }
    return null
  }

  private fun resolvedInset(v: LengthPercentageAuto, dimension: Int): Int? {
    return when (v) {
      is LengthPercentageAuto.Auto -> null
      is LengthPercentageAuto.Zero -> 0
      is LengthPercentageAuto.Points -> v.points.toInt()
      is LengthPercentageAuto.Percent -> (dimension * v.percentage).toInt()
    }
  }

  // Captures a sticky node's natural position, registers it, and returns its current offset.
  fun captureSticky(node: Node, view: View, x: Int, y: Int, width: Int, height: Int): Pair<Int, Int> {
    val host = nearestScrollAncestor(view)
    val parentView = view.parent as? View
    if (host == null || parentView == null) {
      node.stickyScrollHost = null
      node.isStickyEngaged = false
      return x to y
    }

    node.stickyLocalX = x
    node.stickyLocalY = y
    val (naturalX, naturalY) = accumulateOrigin(x, y, view, host)
    node.stickyNaturalX = naturalX
    node.stickyNaturalY = naturalY
    node.stickyWidth = width
    node.stickyHeight = height
    val (cbX, cbY) = accumulateOrigin(parentView.left, parentView.top, parentView, host)
    node.stickyCbLeft = cbX
    node.stickyCbTop = cbY
    node.stickyCbRight = cbX + parentView.width
    node.stickyCbBottom = cbY + parentView.height
    node.stickyScrollHost = host

    host.registerSticky(view)

    val (dx, dy) = stickyDelta(node, host)
    updateEngagement(node, view, dx, dy)
    return (x + dx) to (y + dy)
  }

  // Called on scroll: recomputes every registered sticky descendant's offset.
  fun recomputeSticky(scrollHost: Scroll, descendants: Collection<View>) {
    for (view in descendants.toList()) {
      val element = view as? Element ?: continue
      val node = element.node
      if (node.style.position != Position.Sticky || node.stickyScrollHost !== scrollHost) continue
      val (dx, dy) = stickyDelta(node, scrollHost)
      val newX = node.stickyLocalX + dx
      val newY = node.stickyLocalY + dy
      if (view.left != newX || view.top != newY) {
        view.layout(newX, newY, newX + view.width, newY + view.height)
      }
      updateEngagement(node, view, dx, dy)
    }
  }

  // Shift off natural position to hold the inset, clamped within the containing block.
  private fun stickyDelta(node: Node, host: Scroll): Pair<Int, Int> {
    val viewportLeft = host.scrollX
    val viewportTop = host.scrollY
    val viewportRight = viewportLeft + host.width
    val viewportBottom = viewportTop + host.height
    val naturalX = node.stickyNaturalX
    val naturalY = node.stickyNaturalY
    val size = node.stickyWidth to node.stickyHeight
    val inset = node.style.inset
    val cbHeight = node.stickyCbBottom - node.stickyCbTop
    val cbWidth = node.stickyCbRight - node.stickyCbLeft

    var dy = 0
    val top = resolvedInset(inset.top, cbHeight)
    val bottom = resolvedInset(inset.bottom, cbHeight)
    if (top != null) {
      val minY = viewportTop + top
      if (naturalY < minY) dy = minY - naturalY
    } else if (bottom != null) {
      val maxY = viewportBottom - bottom - size.second
      if (naturalY > maxY) dy = maxY - naturalY
    }
    val minDy = node.stickyCbTop - naturalY
    val maxDy = node.stickyCbBottom - size.second - naturalY
    dy = dy.coerceIn(minOf(minDy, maxDy), maxOf(minDy, maxDy))

    var dx = 0
    val left = resolvedInset(inset.left, cbWidth)
    val right = resolvedInset(inset.right, cbWidth)
    if (left != null) {
      val minX = viewportLeft + left
      if (naturalX < minX) dx = minX - naturalX
    } else if (right != null) {
      val maxX = viewportRight - right - size.first
      if (naturalX > maxX) dx = maxX - naturalX
    }
    val minDx = node.stickyCbLeft - naturalX
    val maxDx = node.stickyCbRight - size.first - naturalX
    dx = dx.coerceIn(minOf(minDx, maxDx), maxOf(minDx, maxDx))

    return dx to dy
  }

  private fun updateEngagement(node: Node, view: View, dx: Int, dy: Int) {
    val engaged = dx != 0 || dy != 0
    if (engaged != node.isStickyEngaged) {
      node.isStickyEngaged = engaged
      if (engaged) {
        (view.parent as? ViewGroup)?.bringChildToFront(view)
      }
    }
  }
}

File: packages/nativescript-masonkit/src-native/mason-ios/Mason/Mason/MasonPositioning.swift
//
//  MasonPositioning.swift
//  Mason
//

import UIKit

enum MasonPositioning {

  // MARK: - Shared

  static func rootView(of node: MasonNode) -> UIView? {
    let root = node.getRootNode()
    if root.type == .document {
      return (root.document?.documentElement as? MasonElement)?.uiView
    }
    return root.view
  }

  // `MasonUIView.isScrollContainer` is true for most plain divs; check the actual overflow instead.
  static func isScrollContainer(_ view: UIView) -> Bool {
    if view is Scroll { return true }
    guard let element = view as? MasonElement else { return false }
    let overflow = element.style.overflow
    return overflow.x == .Scroll || overflow.x == .Auto || overflow.y == .Scroll || overflow.y == .Auto
  }

  // Sum of frame origins from `view` up to `host`; scroll offsets don't affect `frame`, so this ignores scrolling.
  static func accumulateOrigin(_ base: CGPoint, from view: UIView, upTo host: UIView) -> CGPoint {
    var origin = base
    var current = view.superview
    while let cur = current, cur !== host {
      origin.x += cur.frame.origin.x
      origin.y += cur.frame.origin.y
      current = cur.superview
    }
    return origin
  }

  private static func resolvedInset(_ v: MasonLengthPercentageAuto, relativeTo dimension: CGFloat) -> CGFloat? {
    switch v {
    case .Auto: return nil
    case .Zero: return 0
    case .Points(let p): return CGFloat(p)
    case .Percent(let p): return dimension * CGFloat(p) / 100
    }
  }

  // MARK: - Fixed

  // `fixed`'s containing block: the tree root's nearest non-scrolling ancestor.
  static func fixedHost(for node: MasonNode) -> UIView? {
    guard let rootView = rootView(of: node) else { return nil }
    if isScrollContainer(rootView), let superview = rootView.superview {
      return superview
    }
    return rootView
  }

  // Reparents to the fixed host once, then rewrites `frame` from root-relative to the host's space.
  static func applyFixed(node: MasonNode, view: UIView, frame: inout CGRect) {
    guard let host = fixedHost(for: node), let rootView = rootView(of: node) else { return }
    if view.superview !== host {
      if node.fixedOriginalSuperview == nil {
        node.fixedOriginalSuperview = view.superview
      }
      host.addSubview(view)
    }
    host.bringSubviewToFront(view)
    node.fixedRootOrigin = frame.origin
    fixedViews.add(view)
    let origin = accumulateOrigin(rootView.frame.origin, from: rootView, upTo: host)
    frame.origin.x += origin.x
    frame.origin.y += origin.y
  }

  // Views placed by applyFixed, so rootDidMove can find them.
  private static let fixedViews = NSHashTable<UIView>.weakObjects()

  // The host keeps its place when the root moves (e.g. NativeScript lays the page
  // out below the action bar after mason's layout), so re-offset fixed boxes.
  static func rootDidMove(_ root: UIView) {
    guard fixedViews.count > 0 else { return }
    for case let view as UIView in fixedViews.allObjects {
      guard let element = view as? MasonElement, element.node.style.position == .Fixed,
            rootView(of: element.node) === root, let host = view.superview else { continue }
      let origin = accumulateOrigin(root.frame.origin, from: root, upTo: host)
      let target = CGPoint(x: element.node.fixedRootOrigin.x + origin.x, y: element.node.fixedRootOrigin.y + origin.y)
      if view.transform.isIdentity && CATransform3DIsIdentity(view.layer.transform) {
        if view.frame.origin != target { view.frame.origin = target }
      } else {
        let center = CGPoint(x: target.x + view.bounds.width / 2, y: target.y + view.bounds.height / 2)
        if view.center != center { view.center = center }
      }
    }
  }

  // Undoes applyFixed once a node stops being fixed.
  static func clearPositioning(node: MasonNode, view: UIView) {
    node.stickyScrollHost = nil
    fixedViews.remove(view)
    if let original = node.fixedOriginalSuperview {
      node.fixedOriginalSuperview = nil
      if view.superview !== original {
        original.addSubview(view)
      }
    }
  }

  // MARK: - Sticky

  // Walks the native superview chain, not Mason's node-parent chain, which other mounted screens can corrupt.
  private static func nearestScrollAncestor(of view: UIView) -> UIView? {
    var current = view.superview
    while let v = current {
      if isScrollContainer(v) {
        return v
      }
      current = v.superview
    }
    return nil
  }

  // Captures a sticky node's natural position, registers it, and applies its current offset.
  static func captureSticky(node: MasonNode, view: UIView, frame: inout CGRect) {
    guard let host = nearestScrollAncestor(of: view), let parentView = view.superview else {
      node.stickyScrollHost = nil
      node.isStickyEngaged = false
      return
    }

    node.stickyLocalOrigin = frame.origin
    node.stickyNaturalOrigin = accumulateOrigin(frame.origin, from: view, upTo: host)
    node.stickyViewSize = frame.size
    node.stickyContainingBlockRect = CGRect(
      origin: accumulateOrigin(parentView.frame.origin, from: parentView, upTo: host),
      size: parentView.bounds.size
    )
    node.stickyScrollHost = host

    if let scroll = host as? Scroll {
      scroll.registerSticky(view)
    } else if let mv = host as? MasonUIView {
      mv.registerSticky(view)
    }

    let delta = stickyDelta(node: node, host: host)
    frame.origin.x += delta.x
    frame.origin.y += delta.y
    updateEngagement(node: node, view: view, delta: delta)
  }

  // Called on scroll: recomputes every registered sticky descendant's offset.
  static func recomputeSticky(scrollHost: UIView, descendants: NSHashTable<UIView>) {
    for case let view as UIView in descendants.allObjects {
      guard let element = view as? MasonElement else { continue }
      let node = element.node
      guard node.style.position == .Sticky, node.stickyScrollHost === scrollHost else { continue }
      let delta = stickyDelta(node: node, host: scrollHost)
      let newOrigin = CGPoint(x: node.stickyLocalOrigin.x + delta.x, y: node.stickyLocalOrigin.y + delta.y)
      if view.frame.origin != newOrigin {
        view.frame.origin = newOrigin
      }
      updateEngagement(node: node, view: view, delta: delta)
    }
  }

  // Shift off natural position to hold the inset, clamped within the containing block.
  private static func stickyDelta(node: MasonNode, host: UIView) -> CGPoint {
    let viewport = host.bounds
    let natural = node.stickyNaturalOrigin
    let size = node.stickyViewSize
    let cb = node.stickyContainingBlockRect
    let inset = node.style.inset

    var dy: CGFloat = 0
    if let top = resolvedInset(inset.top, relativeTo: cb.height) {
      let minY = viewport.origin.y + top
      if natural.y < minY { dy = minY - natural.y }
    } else if let bottom = resolvedInset(inset.bottom, relativeTo: cb.height) {
      let maxY = viewport.origin.y + viewport.height - bottom - size.height
      if natural.y > maxY { dy = maxY - natural.y }
    }
    let minDy = cb.minY - natural.y
    let maxDy = cb.maxY - size.height - natural.y
    dy = max(minDy, min(dy, maxDy))

    var dx: CGFloat = 0
    if let left = resolvedInset(inset.left, relativeTo: cb.width) {
      let minX = viewport.origin.x + left
      if natural.x < minX { dx = minX - natural.x }
    } else if let right = resolvedInset(inset.right, relativeTo: cb.width) {
      let maxX = viewport.origin.x + viewport.width - right - size.width
      if natural.x > maxX { dx = maxX - natural.x }
    }
    let minDx = cb.minX - natural.x
    let maxDx = cb.maxX - size.width - natural.x
    dx = max(minDx, min(dx, maxDx))

    return CGPoint(x: dx, y: dy)
  }

  private static func updateEngagement(node: MasonNode, view: UIView, delta: CGPoint) {
    let engaged = delta.x != 0 || delta.y != 0
    if engaged != node.isStickyEngaged {
      node.isStickyEngaged = engaged
      if engaged, let superview = view.superview {
        superview.bringSubviewToFront(view)
      }
    }
  }
}
